File: packages/sdk/src/resources/web-bot-auth.ts
import type { LinkOptions } from '@/config';
import { LinkSdkError } from '@/errors';
import { BaseResource, requireRecord, requireString } from '@/resources/base';
import type { IWebBotAuthResource } from '@/resources/interfaces';
import type { WebBotAuthBlock } from '@/types/index';

interface CacheEntry {
  block: WebBotAuthBlock;
  expiresAt: number;
}

const EXPIRY_BUFFER_MS = 30_000;

function parseWebBotAuthBlock(value: unknown): WebBotAuthBlock {
  const body = requireRecord(value);
  return {
    signature: requireString(body.signature, 'web_bot_auth.signature'),
    signature_input: requireString(
      body.signature_input,
      'web_bot_auth.signature_input',
    ),
    signature_agent: requireString(
      body.signature_agent,
      'web_bot_auth.signature_agent',
    ),
    authority: requireString(body.authority, 'web_bot_auth.authority'),
    expires_at: requireString(body.expires_at, 'web_bot_auth.expires_at'),
  };
}

export class WebBotAuthResource
  extends BaseResource
  implements IWebBotAuthResource
{
  private readonly cache = new Map<string, CacheEntry>();

  constructor(options: LinkOptions) {
    super(options, '/web_bot_auth/sign');
  }

  async signUrl(url: string): Promise<WebBotAuthBlock> {
    let authority: string;
    try {
      authority = new URL(url).hostname;
    } catch (error) {
      throw new LinkSdkError(`Invalid URL: ${url}`, { cause: error });
    }

    const cached = this.cache.get(authority);
    if (cached && Date.now() < cached.expiresAt - EXPIRY_BUFFER_MS) {
      return cached.block;
    }

    const { status, data, rawBody } = await this.apiFetch({
      method: 'POST',
      url: this.endpoint,
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ url }),
    });

    if (status < 200 || status >= 300) {
      this.throwApiError('get web bot auth headers', status, data, rawBody);
    }

    const webBotAuth = this.parseResponse(
      'get web bot auth headers',
      status,
      () => {
        const body = requireRecord(data);
        return parseWebBotAuthBlock(body.web_bot_auth);
      },
    );
    const expiresAt = Date.parse(webBotAuth.expires_at);
    if (Number.isNaN(expiresAt)) {
      throw new LinkSdkError(
        `Credentials response has invalid expires_at: ${webBotAuth.expires_at}`,
      );
    }

    this.cache.set(authority, { block: webBotAuth, expiresAt });
    return webBotAuth;
  }
}

File: packages/sdk-go/web_bot_auth.go
package link

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"net/url"
	"sync"
	"time"
)

const webBotAuthExpiryBuffer = 30 * time.Second

type webBotAuthCacheEntry struct {
	block     WebBotAuthBlock
	expiresAt time.Time
}

// WebBotAuthResource signs merchant URLs for Web Bot Auth.
type WebBotAuthResource struct {
	base  *baseResource
	mu    sync.RWMutex
	cache map[string]webBotAuthCacheEntry
	now   func() time.Time
}

func newWebBotAuthResource(base *baseResource) *WebBotAuthResource {
	return &WebBotAuthResource{
		base:  base,
		cache: make(map[string]webBotAuthCacheEntry),
		now:   time.Now,
	}
}

// SignURL returns signing material for the URL's authority. Fresh credentials
// are cached by hostname until 30 seconds before expiration.
func (r *WebBotAuthResource) SignURL(ctx context.Context, rawURL string) (*WebBotAuthBlock, error) {
	parsed, err := url.Parse(rawURL)
	if err != nil || parsed.Scheme == "" || parsed.Hostname() == "" {
		if err == nil {
			err = fmt.Errorf("URL must be absolute")
		}
		return nil, newSDKError("Invalid URL: "+rawURL, err)
	}
	authority := parsed.Hostname()

	r.mu.RLock()
	cached, ok := r.cache[authority]
	r.mu.RUnlock()
	if ok && r.now().Before(cached.expiresAt.Add(-webBotAuthExpiryBuffer)) {
		block := cached.block
		return &block, nil
	}

	var envelope struct {
		WebBotAuth WebBotAuthBlock `json:"web_bot_auth"`
	}
	requestBody, err := json.Marshal(struct {
		URL string `json:"url"`
	}{URL: rawURL})
	if err != nil {
		return nil, err
	}
	response, err := r.base.fetch(
		ctx,
		http.MethodPost,
		r.base.baseURL+"/web_bot_auth/sign",
		http.Header{"Content-Type": []string{"application/json"}},
		requestBody,
	)
	if err != nil {
		return nil, err
	}
	if response.status < 200 || response.status >= 300 {
		return nil, newAPIError("get web bot auth headers", response.status, response.data, response.rawBody)
	}
	if err := decodeResponse("get web bot auth headers", response, &envelope); err != nil {
		return nil, err
	}
	block := envelope.WebBotAuth
	if block.Signature == "" || block.SignatureInput == "" || block.SignatureAgent == "" || block.Authority == "" || block.ExpiresAt == "" {
		return nil, newResponseError("get web bot auth headers", response.status, errors.New("response is missing required signing fields"))
	}
	expiresAt, err := time.Parse(time.RFC3339, block.ExpiresAt)
	if err != nil {
		return nil, newResponseError("get web bot auth headers", response.status, fmt.Errorf("invalid expires_at %q: %w", block.ExpiresAt, err))
	}
	r.mu.Lock()
	r.cache[authority] = webBotAuthCacheEntry{block: block, expiresAt: expiresAt}
	r.mu.Unlock()
	return &envelope.WebBotAuth, nil
}
